packages/lesswrong/server/scripts/migrateCollections.ts

```diff
@@ -0,0 +1,181 @@
+
+const copyData = async (
+  sql: Transaction,
+  collections: SwitchingCollection<DbObject>[],
+  pass: number,
+  createdSince?: Date,
+) => {
+  console.log(`...Copying data (pass ${pass})`);
+  for (const collection of collections) {
+    console.log(`......${collection.getName()}`);
+    const table = collection.getPgCollection().table;
+
+    const formatter = getCollectionFormatter(collection);
+    const batchSize = pickBatchSize(collection);
+    let count = 0;
+    await forEachDocumentBatchInCollection({
+      collection: collection.getMongoCollection() as unknown as CollectionBase<DbObject>,
+      batchSize,
+      filter: makeBatchFilter(createdSince),
+      callback: async (documents: DbObject[]) => {
+        console.log(`.........Migrating ${collection.getName()} documents ${count}-${count + documents.length}`);
+        count += batchSize;
+        const query = new InsertQuery(table, documents.map(formatter), {}, {conflictStrategy: "ignore"});
+        const compiled = query.compile();
+        try {
+          await sql.none(compiled.sql, compiled.args);
+        } catch (e) {
+          throw new Error(`Error importing document batch for collection ${collection.getName()}: ${e.message}`);
+        }
+      },
+    });
+  }
+}
+
+export const migrateCollections = async (collectionNames: CollectionNameString[]) => {
+  console.log(`=== Migrating collections '${collectionNames}' from Mongo to Postgres ===`);
+
+  console.log("...Preparing collections");
+  const collections = collectionNames.map(getCollection) as unknown as SwitchingCollection<DbObject>[];
+  for (const [index, collection] of collections.entries()) {
+    if (!collection) {
+      throw new Error(`Invalid collection name '${collectionNames[index]}'`);
+    }
+
+    if (!(collection instanceof SwitchingCollection)) {
+      throw new Error(`Collection '${collectionNames[index]}' is not a SwitchingCollection`);
+    }
+
+    collection.setTargets("mongo", "mongo");
+  }
+
+  console.log("...Beginning SQL transaction");
+  const sql = getSqlClientOrThrow();
+  try {
+    await sql.tx(async (transaction) => {
+      // Create the tables
+      await createTables(transaction, collections);
+      await createIndexes(transaction, collections);
+      await setLogged(transaction, collections, false);
+
+      // Copy the initial data - this can take a *long* time depending on the collection
+      const copyStart = new Date();
+      await copyData(transaction, collections, 1);
+      await setLogged(transaction, collections, true);
+
+      // Copy anything that was inserted during the first copy - this should be ~instant
+      const copyStart2 = new Date();
+      await copyData(transaction, collections, 2, copyStart);
+
+      // Start writing to Postgres
+      for (const collection of collections) {
+        collection.setTargets("mongo", "pg");
+      }
+
+      // Once more for luck...
+      await copyData(transaction, collections, 3, copyStart2);
```

Comment:
nice

Reply:
Although won't the other servers still be writing to mongo at this point so some stuff could end up not being copied over? I think the process should be:
 - first two data copies
 - disable writes to mongo (for all servers, wait for some signal via the lock table that all servers have done this)
 - third data copy
 - enable reads and writes to postgres (for all servers, again wait for signal via the lock table that all servers have done this successfully)

I guess having everything in a transaction makes it hard to communicate via the lock table, could the lock table be a mongo collection instead?

Reply:
Ah, yeah - you're right. That _is_ the process that we're using right now (when we first set the write target to "pg", that implicitly disables mongo writes), but, as you point out, we only do that on the machine doing the copying, not the server instances, so we need to propogate that through the lock table.

We can still execute SQL queries here without worrying about the transaction - we just have to use `sql` instead of `transaction`.

The main question is, is it sufficient to just write to the lock table and then wait a certain period of time (3 * POLL_RATE_SECONDS maybe?) and trust that the other instances switched, or do we really need two-way communication through the lock? If we do need two-way comms then that's a lot more complicated - I think the simplest way is to have some counter that we can atomically increment, but then we still need to hard code the number of instances which is annoying, and I'm pretty sure it's different for LessWrong.